contracts/libs/0.8.x/BytecodeStorageV1.sol

```diff
@@ -187,20 +228,117 @@ library BytecodeStorage {
 
     /**
-        @notice Returns the size of the bytecode at address `_address`
-        @param _address address that may or may not contain bytecode
-        @return size size of the bytecode code at `_address`
-    */
+     * @notice Returns the size of the bytecode at address `_address`
+     * @param _address address that may or may not contain bytecode
+     * @return size size of the bytecode code at `_address`
+     */
     function _bytecodeSizeAt(
         address _address
     ) private view returns (uint256 size) {
         assembly {
             size := extcodesize(_address)
         }
     }
+
+    /**
+     * @notice Returns the offset of the data in the bytecode at address `_address`
+     * @param _address address that may or may not contain bytecode
+     * @return dataOffset offset of data in bytecode if a known version, otherwise 0
+     */
+    function _bytecodeDataOffsetAt(
+        address _address
+    ) private view returns (uint256 dataOffset) {
+        bytes32 version = _bytecodeVersionAt(_address);
+        if (version == V1_VERSION_STRING) {
+            dataOffset = V1_DATA_OFFSET;
+        } else if (version == V0_VERSION_STRING) {
+            dataOffset = V0_DATA_OFFSET;
+        } else {
+            dataOffset = 0;
+        }
+    }
+
+    /**
+     * @notice Returns the offset of the address in the bytecode at address `_address`
+     * @param _address address that may or may not contain bytecode
+     * @return addressOffset offset of address in bytecode if a known version, otherwise 0
+     */
+    function _bytecodeAddressOffsetAt(
+        address _address
+    ) private view returns (uint256 addressOffset) {
+        bytes32 version = _bytecodeVersionAt(_address);
+        if (version == V1_VERSION_STRING) {
+            addressOffset = V1_ADDRESS_OFFSET;
+        } else if (version == V0_VERSION_STRING) {
+            addressOffset = V0_ADDRESS_OFFSET;
+        } else {
+            addressOffset = 0;
+        }
+    }
+
+    /**
+     * @notice Get version string for given contract bytecode
+     * @param _address address of deployed contract with bytecode containing concat(version, deployer-address, data)
+     * @return version version string read from contract bytecode
+     */
+    function _bytecodeVersionAt(
+        address _address
+    ) private view returns (bytes32 version) {
+        // get the size of the data
+        uint256 bytecodeSize = _bytecodeSizeAt(_address);
+        // handle case where address contains code < VERSION_OFFSET
+        // note: the first check here also captures the case where
+        //       (bytecodeSize == 0) implicitly, but we add the second check of
+        //       (bytecodeSize == 0) as a fall-through that will never execute
+        //       unless `VERSION_OFFSET` is set to 0 at some point.
+        if ((bytecodeSize < VERSION_OFFSET) || (bytecodeSize == 0)) {
+            revert("ContractAsStorage: Read Error");
+        }
```

Comment:
I think subcall in `_bytecodeSizeAt` reverts if `bytecodeSize == 0`, so can probably drop that here?
Also, do we want to check instead for the case `bytecodeSize < VERSION_OFFSET + 32`?

Reply:
Updated!